Add glib event loop

jkonecny12 opened the pull request:
Most of the related old tests are also used GLib event loop.

This event loop is not required by the Simpleline package. You do not need to use it and if you want to use it, please add it as requires for your package / install it manually.

poncovka commented on tests/__init__.py:
```diff
@@ -20,23 +20,22 @@
-    separator = "\n".join(2 * [width * "="])
-    separator += "\n"  # print adds another newline
-    return separator
+class UtilityFunctions(object):
```
Maybe `UtilityMixin`?

jkonecny12 replied:
Good idea, I like that.

poncovka commented on tests/glib_tests/__init__.py:
```diff
@@ -17,41 +17,47 @@
 
-class GLibProcessEvents_TestCase(ProcessEvents_TestCase):
-    """Run all the tests in ProcessEvents test case but with GLib event loop."""
+class GLibUtilityFunctions(object):
```
`GLibUtilityMixin`?

jkonecny12 replied:
I like that.

poncovka commented on simpleline/event_loop/main_loop.py:
```diff
@@ -161,19 +163,35 @@ def process_signals(self, return_after=None):
+                # Signal with this priority is not available anymore
+                if signal is None:
+                    return
+                priority = signal.priority
```
There is no reason to combine the two ways how to process signals. You ask again and again if `return_after` or `priority` is (not) None in a potentially frequently run place. You could split this code to two branches the same way as in `GLibEventLoop`. It means that you can also remove the method `_take_signal_for_processing` and remove `wait_on_signal `from `_check_if_signal_processed`.

jkonecny12 replied:
Fixed.

jkonecny12 commented on simpleline/event_loop/event_queue.py:
```diff
@@ -94,6 +94,22 @@ def get(self):
         """
         return self._queue.get()
 
+    def get_top_event_when_priority(self, priority):
```
Sounds better but it is not true. You will take the event with this priority only if this event is on top. So when you ask for lower priority event which is event queue however it is not on a top you won't get it.

Changing to `get_top_event_if_priority()`.

jkonecny12 commented:
**UPDATED:**
* Rebased on commits from PR https://github.com/rhinstaller/python-simpleline/pull/37 .
* Fixed issues in PR https://github.com/rhinstaller/python-simpleline/pull/37 .
* Fixed issues here.

jkonecny12 commented:
**UPDATED:**
* Remove the `_take_signal_for_processing()` method. I must admit, it is much better now :)

jkonecny12 commented:
**UPDATED:**
* Added force quit feature.

jkonecny12 commented:
**UPDATED:**
I finally fixed all the issues I was able to find. Many minor changes here.

poncovka commented on simpleline/event_loop/main_loop.py:
```diff
@@ -161,135 +155,96 @@ def _mainloop(self):
             self._process_signal(signal)
 
+            # all who is waiting on this signal can stop waiting
+            self._mark_signal_processed(signal)
```
You are marking the signal twice. Is it alright?

jkonecny12 replied:
It is not problem, however, it can be removed.
Will be fixed.

poncovka commented on simpleline/event_loop/main_loop.py:
```diff
@@ -161,135 +155,96 @@ def _mainloop(self):
+            if signal is None:
+                return
+
+            priority = signal.priority
```
`pritority` could be set in the `if priority is None` branch, because it cannot change during the iteration, right?

jkonecny12 replied:
Yeah it could but why? I see it as a pointless line .

jkonecny12 replied:
Fixed.

jkonecny12 commented:
**UPDATED:**
* Tweak based on @poncovka's suggestions.
